community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/commands/predicates/ConstantCachedIn.scala

```diff
@@ -0,0 +1,133 @@
+The key for the cache is the expression and not the value, which saves in execution speed
+ */
+case class ConstantCachedIn(value: Expression, list: Expression) extends Predicate with CollectionSupport {
+
+  // These two are here to make the fields accessible without conflicting with the case classes
+  override def isMatch(ctx: ExecutionContext)(implicit state: QueryState) = {
+    val inChecker = state.cachedIn.getOrElseUpdate(list, {
+      val listValue = list(ctx)
+      val checker = if (listValue == null)
+        NullListChecker
+      else {
+        val input = makeTraversable(listValue).toIterator
+        if (input.isEmpty) AlwaysFalseChecker else new BuildUp(input)
+      }
+      new InCheckContainer(checker)
+    })
+
+    inChecker.contains(value(ctx))
+  }
+
+  override def containsIsNull = false
+
+  override def arguments = Seq(list)
+
+  override def symbolTableDependencies = list.symbolTableDependencies ++ value.symbolTableDependencies
+
+  override def rewrite(f: (Expression) => Expression) = f(ConstantCachedIn(value.rewrite(f), list.rewrite(f)))
+}
+
+/*
+This class is used for making the common <exp> IN <rhs-expression> fast
+
+It uses a cache for the <rhs-expression> value, and turns it into a Set, for fast existence checking
+ */
+case class DynamicCachedIn(value: Expression, list: Expression) extends Predicate with CollectionSupport {
+
+  // These two are here to make the fields accessible without conflicting with the case classes
+  override def isMatch(ctx: ExecutionContext)(implicit state: QueryState): Option[Boolean] = {
+    val listValue = list(ctx)
+
+    if(listValue == null)
+      return None
+
+    val traversable = makeTraversable(listValue)
+    val inputIterator = traversable.toIterator
+
+    if(inputIterator.isEmpty)
+      return Some(false)
+
+    val inChecker = state.cachedIn.getOrElseUpdate(traversable, {
+      val checker = new BuildUp(inputIterator)
+      new InCheckContainer(checker)
+    })
+
+    inChecker.contains(value(ctx))
+  }
+
+  override def containsIsNull = false
+
+  override def arguments = Seq(list)
+
+  override def symbolTableDependencies = list.symbolTableDependencies ++ value.symbolTableDependencies
+
+  override def rewrite(f: (Expression) => Expression) = f(DynamicCachedIn(value.rewrite(f), list.rewrite(f)))
+}
+
+object CachedIn {
+  def unapply(arg: Expression): Option[(Expression, Expression)] = arg match {
+    case DynamicCachedIn(value, list) => Some((value, list))
+    case ConstantCachedIn(value, list) => Some((value, list))
+    case _ => None
+  }
+}
+
+/*
+This is a simple container that keep the latest state of the cached IN check
+ */
+class InCheckContainer(var checker: Checker) {
+  def contains(value: Any): Option[Boolean] = {
+    val (result, newChecker) = checker.contains(value)
+    checker = newChecker
+    result
+  }
+}
+
+class SingleThreadedLRUCache[K, V](maxSize: Int) {
+  val cache: ArrayBuffer[(K, V)] = new ArrayBuffer[(K, V)](maxSize)
```

Comment:
it's done this way because we want to avoid calculating hashCodes. The default is to keep only 16 slots of cache, so in this case, linear scanning an arraay is faster than a hash map would be.